File: examples/15_algorithms/15a_nqueens/nqueens.c
/* The same problem as nqueens.ppy, reading the judge's input with scanf. */
#include <stdio.h>

static long long solve(long long full, long long columns, long long diagonal,
                       long long antidiagonal) {
    if (columns == full) {
        return 1;
    }
    long long total = 0;
    long long available = full & ~(columns | diagonal | antidiagonal);
    while (available != 0) {
        long long bit = available & -available;
        available -= bit;
        total += solve(full, columns | bit, ((diagonal | bit) << 1) & full,
                       (antidiagonal | bit) >> 1);
    }
    return total;
}

int main(void) {
    int n = 0;
    if (scanf("%d", &n) != 1) {
        n = 12;
    }

    long long answer = solve((1LL << n) - 1, 0, 0, 0);

    printf("%lld\n", answer);
    return 0;
}

File: examples/15_algorithms/15a_nqueens/nqueens.py
"""N-Queens: how many ways N queens fit on an N by N board.

Input:  N on the first line.
Output: how many ways N queens can be placed without attacking each other.
"""


def solve(full, columns, diagonal, antidiagonal):
    if columns == full:
        return 1
    total = 0
    available = full & ~(columns | diagonal | antidiagonal)
    while available != 0:
        bit = available & -available
        available -= bit
        total += solve(
            full,
            columns | bit,
            ((diagonal | bit) << 1) & full,
            (antidiagonal | bit) >> 1,
        )
    return total


def read_size():
    """The judge's input, or the stand-in this example uses on its own."""
    try:
        return int(input())
    except EOFError:
        return 12


def main():
    n = read_size()
    print(solve((1 << n) - 1, 0, 0, 0))


main()
